go/stats/metrics.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"math"
+	"net"
+	"strconv"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/inverse-inc/packetfence/go/interval"
+	"github.com/inverse-inc/packetfence/go/log"
+	"github.com/inverse-inc/packetfence/go/pfconfigdriver"
+	util "github.com/inverse-inc/packetfence/go/sharedutils"
+)
+
+var MySQLdatabase *sql.DB
+
+func ProcessMetricConfig(ctx context.Context, conf pfconfigdriver.PfStats) error {
+
+	job := func() {}
+
+	switch conf.Type {
+	case "mysql_query":
+		job = func() {
+			rows, err := MySQLdatabase.Query(conf.MySQLQuery)
+			if err != nil {
+				log.LoggerWContext(ctx).Error(err.Error())
+				return
+			}
+			defer rows.Close()
+			var (
+				result string
+			)
+			for rows.Next() {
+				err := rows.Scan(&result)
+				if err != nil {
+					log.LoggerWContext(ctx).Error(err.Error())
+					return
+				}
+				switch conf.StatsdType {
```

Comment:
Last detail before we merge is to get rid of that duplicated switch and extract that into a function so we can reuse it in all the metrics without copy/pasting it

Reply:
@julsemaan `increment` and `unique` are useless in `icmp_ipv4`, while `timing` is only useful in `icmp_ipv4`. Since the switch differs between each of their use-cases, they can't be re-used.

Reply:
In that case, we could still create the function but add a validation through a map that ensures you only use valid statsd types for a specific metric type

Reply:
@julsemaan it doesn't seem counter-intuitive to always have at least one function parameter that isn't used? In order to create a function I'd need to utilize both a `metric interface{}` and a `client StatsdClient` parameter, and in every case at least one is not used due to:

```
					case "timing":
						client.Send(conf.StatsdNS)
```

Reply:
Not sure I fully followed you on that one but here is the pseudo-code of what I had in mind:

```
func recordMetric(conf pfconfigdriver.PfStats, stat float64) error {
    // That should be a constant
    valid := map[string]map[string]bool{
        "mysql_query":map[string]bool{
              "count":true,
              "gauge":true,
        }
    }

    if validTypes, ok := valid[conf.Type]; !ok {
        return errors.New("Unrecognized metric type")
    } else {
        if _, ok := validTypes[conf.StatsdType]; !ok {
            return errors.New("Invalid statsd type")
        }
    }

    switch {
        .........
    }
}
```

With that above, we would just have to call the function and it would handle validating the statsd type works with the metric type and would centralize the whole logic of sending a metric to statsd.

Reply:
@julsemaan the above won't work. See line line 112 where the `timing` uses the `Statsdclient` type and not a ` float`.

Reply:
Instead of using a wrapper that automatically sends the timing in that case, we could just track the start and the finish of the ping and then send it to StatsdClient.Timing("whatever", timing)

Reply:
@julsemaan this would prevent being able to use any of the other chart types for a timing. I abstracted it this way so the user could decide to use either a `count`, `gauge`, `histogram` or the normal `timing` chart for `icmp`.

Reply:
Ah, yes I do see this now.

IMO, I'd have kept ICMP to a simple timing because with every timing metric, you automatically get a count in statsd, but then, the code is there so we can leave it but it might be redundant a little bit for that part but to get a histogram we would need the implementation you did. We don't really use them but lets leave it as is in case we ever want to use them

Reply:
@satkunas 

@extrafu can confirm but I believe we do want this.

So as discussed, I think the best route would be to code it instead of doing it via stats.conf